### PR TITLE
C++: Taint tests for array and reference assignments.

### PR DESCRIPTION
**Pull request by geoffw0**
I wrote these tests a few days ago to better understand where we do and don't follow taint through references and pointers.  I think we might as well add them to our battery of tests.  They highlight some differences between our IR and AST taint flow.

### CODE REVIEW

File: cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+};
+
+void test_myint_member_assignment()
+{
+	MyInt mi;
+
+	mi.i = source();
+
+	sink(mi); // tainted [DETECTED BY IR ONLY]
+	sink(mi.get()); // tainted
```
Comment (MathiasVP) [inline cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:57]: This one is not detected by IR, right?

File: cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+void test_myint_method_assignment()
+{
+	MyInt mi;
+
+	mi.get() = source();
+
+	sink(mi); // tainted [DETECTED BY IR ONLY]
+	sink(mi.get()); // tainted
```
Comment (MathiasVP) [inline cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:67]: Also not detected by IR (unless I'm reading the .expected files incorrectly).

**Review by MathiasVP — COMMENTED** [review]
LGTM otherwise!

File: cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:67]: The two results appear to be spotted by both, in slighty differing forms.  In `taint.expected` (the AST taint results):
```
| arrayassignment.cpp:57:10:57:12 | call to get | arrayassignment.cpp:54:9:54:14 | call to source |
| arrayassignment.cpp:67:10:67:12 | call to get | arrayassignment.cpp:64:13:64:18 | call to source |
```
In `test_ir.expected` (the IR taint results):
```
| arrayassignment.cpp:56:7:56:8 | mi | arrayassignment.cpp:54:9:54:14 | call to source |
| arrayassignment.cpp:57:10:57:15 | (reference dereference) | arrayassignment.cpp:54:9:54:14 | call to source |
| arrayassignment.cpp:66:7:66:8 | mi | arrayassignment.cpp:64:13:64:18 | call to source |
| arrayassignment.cpp:67:10:67:15 | (reference dereference) | arrayassignment.cpp:64:13:64:18 | call to source |
```

Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/arrayassignment.cpp:67]: Ah! Thanks for clearing that up.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**